torch/nn/attention/_fa4.py

```diff
@@ -0,0 +1,599 @@
+# mypy: allow-untyped-defs
+
+from __future__ import annotations
+
+import importlib
+from dataclasses import dataclass
+from typing import Any, TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from types import ModuleType
+
+import torch
+from torch.library import Library
+
+
+__all__ = [
+    "register_flash_attention_fa4",
+    "flash_attention_fa4_status",
+]
+
+
+@dataclass
+class _FA4FlashAttentionState:
+    module_path: str = "flash_attn.cute.interface"
+    module: ModuleType | None = None
+    registered: bool = False
+    last_error: str | None = None
+
+
+_FA4_STATE = _FA4FlashAttentionState()
+_FA4_LIBRARY: Library | None = None
+
+
+def _flash_attention_forward_fallback(*args, **kwargs):
+    return torch.ops.aten._flash_attention_forward.default(*args, **kwargs)
+
+
+def _flash_attention_backward_fallback(*args, **kwargs):
+    return torch.ops.aten._flash_attention_backward.default(*args, **kwargs)
+
+
+def _sdpa_forward_fallback(*args, **kwargs):
+    return torch.ops.aten._scaled_dot_product_flash_attention.default(*args, **kwargs)
+
+
+def _sdpa_backward_fallback(*args, **kwargs):
+    return torch.ops.aten._scaled_dot_product_flash_attention_backward.default(
+        *args, **kwargs
+    )
+
+
+def register_flash_attention_fa4(
+    module_path: str = "flash_attn.cute.interface",
+) -> None:
+    module = _fa4_import_module(module_path)
+    _FA4_STATE.module_path = module_path
+    _FA4_STATE.module = module
+    _FA4_STATE.last_error = None
+    if _FA4_STATE.registered:
+        return
+    _fa4_register_kernels()
+    _FA4_STATE.registered = True
+
+
+def flash_attention_fa4_status() -> dict[str, Any]:
+    return {
+        "registered": _FA4_STATE.registered,
+        "module_loaded": _FA4_STATE.module is not None,
+        "module_path": _FA4_STATE.module_path,
+        "last_error": _FA4_STATE.last_error,
+    }
+
+
+def _fa4_import_module(module_path: str) -> ModuleType:
+    spec = importlib.util.find_spec(module_path)
+    if spec is None:
+        message = f"Module '{module_path}' not found"
+        _FA4_STATE.last_error = message
+        raise ImportError(message)
+    module = importlib.import_module(module_path)
+    if not hasattr(module, "_flash_attn_fwd") or not hasattr(module, "_flash_attn_bwd"):
+        message = f"Module '{module_path}' does not expose FA4 kernels"
+        _FA4_STATE.last_error = message
+        raise RuntimeError(message)
```

Comment:
I'm a bit curious why this is not literraly doing `import flash_attn.cute.interface` ?